Re-enables paxtest in testinfra

conorsch opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes #1039 

These tests were skipped because we weren't installing paxtest by
default. Let's just install it as part of the test run. Removes
the skip-in-prod marker, so that these checks can be used as part of QA
on hardware. Updates the logic to be distro-specific.

## Testing

CI should be passing on both platforms.
Try to run against hardware, or at least prodVMs, via the `./securedrop-admin verify` workflow.
Are we OK with installing `paxtest` automatically via tests? I'd think so, given that we do it manually as part of QA anyway.

## Deployment
Mostly QA, so developer-focused. Technically if an Admin runs `./securedrop-admin verify`, the remote state will change because `paxtest` will be installed now. That's acceptable, IMO, since otherwise the paxtest checks would have been skipped. 

## Checklist

Commit:
Re-enables paxtest in testinfra

These tests were skipped because we weren't installing paxtest by
default. Let's just install it as part of the test run. Removes
the skip-in-prod marker, so that these checks can be used as part of QA
on hardware. Updates the logic to be distro-specific.

Removes paxtest after test run

We don't install it by default, and it's only useful in QA, so let's
have the test-only dependency automatically cleaned up after install.

zenmonkeykstop commented:
In the spirit of "leave no trace," the script wrapping pytest (`devops/scripts/run_prod_testinfra`) has a cleanup trap function - could stick some apt removes in there.

eloquence commented:
Seemed to work as expected in https://github.com/freedomofpress/securedrop/issues/5609#issuecomment-799834422, agree that it'd be nice to automatically uninstall paxtest in the cleanup function after it's done.

conorsch commented:
Rebased on latest develop. I added a commit to clean up the package afterwards. Didn't use the bash trap, since that required futzing with Ansible outside of our wrappers within the Tails environment. Instead, I just used a try/finally block in the test itself. End result is the same: after test run, regardless of whether the paxtest failed or not, the package will be removed.

Ready for re-review.

eloquence commented:
(Kicking CI)

conorsch commented:
Rebased on top of latest develop, to satisfy CI.

zenmonkeykstop commented:
Fails on a fresh install, as the apt cache has not been updated with `universe`. After the first run of `apt-get update`, paxtest is available as an option, and the tests pass.

zenmonkeykstop commented on molecule/testinfra/common/test_grsecurity.py:
```diff
@@ -99,39 +103,46 @@ def test_grsecurity_sysctl_options(host, sysctl_opt):
+    if not host.exists("/usr/bin/paxtest"):
+        warnings.warn("Installing paxtest to run kernel tests")
+        with host.sudo():
+            host.run("apt-get install -y paxtest")
```
Assumes that paxtest is available via apt - which may not be the case until `apt-get update` has run at least once on the host. Running `apt get update && apt-get install -y paxtest` would work.

conorsch replied:
Added!

zenmonkeykstop commented on molecule/testinfra/common/test_grsecurity.py:
```diff
@@ -99,39 +103,46 @@ def test_grsecurity_sysctl_options(host, sysctl_opt):
+                                                 paxtest_results.split('\n')):
+            print(paxtest_diff)
+        assert paxtest_results == paxtest_expected
+    finally:
```
Paxtest isn't removed in this finally clause, it needs `sudo`.

conorsch replied:
Good catch! Fixed.

zenmonkeykstop reviewed (changes requested):
One final tweak needed to remove the paxtest package post-run...
